crates/wassette/src/config.rs

```diff
@@ -0,0 +1,239 @@
+        self
+    }
+
+    /// Override the OCI client.
+    pub fn with_oci_client(mut self, client: oci_client::Client) -> Self {
+        self.oci_client = Some(client);
+        self
+    }
+
+    /// Control whether the manager eagerly loads components during build.
+    pub fn with_eager_loading(mut self, eager: bool) -> Self {
+        self.eager_load = eager;
+        self
+    }
+
+    /// Produce a validated [`LifecycleConfig`] without constructing a manager.
+    pub fn build_config(self) -> Result<LifecycleConfig> {
+        let plugin_dir = match self.plugin_dir.canonicalize() {
+            Ok(path) => path,
+            Err(_) => self.plugin_dir.clone(),
+        };
+
+        let secrets_dir = self.secrets_dir.unwrap_or_else(get_default_secrets_dir);
+
+        let http_client = match self.http_client {
+            Some(client) => client,
+            None => default_http_client()?,
+        };
+
+        let oci_client = match self.oci_client {
+            Some(client) => client,
+            None => default_oci_client()?,
+        };
+
+        Ok(LifecycleConfig {
+            plugin_dir,
+            secrets_dir,
+            environment_vars: self.environment_vars,
+            http_client,
+            oci_client,
+            eager_load: self.eager_load,
+        })
+    }
+
+    /// Construct a [`LifecycleManager`] using the current builder settings.
+    ///
+    /// If eager loading is enabled the plugin directory is scanned
+    /// immediately; otherwise the caller can defer loading until a later
+    /// [`LifecycleManager::load_all_components`](crate::LifecycleManager::load_all_components)
+    /// invocation.
+    pub async fn build(self) -> Result<LifecycleManager> {
+        let config = self.build_config()?;
+        let eager = config.eager_load();
+        let manager = LifecycleManager::from_config(config).await?;
+        if eager {
+            manager.load_all_components().await?;
+        }
+        Ok(manager)
+    }
+}
+
+/// Create the default HTTP client used when none is supplied.
+fn default_http_client() -> Result<reqwest::Client> {
+    let http_timeout = std::env::var("HTTP_TIMEOUT_SECS")
+        .ok()
+        .and_then(|s| s.parse().ok())
+        .unwrap_or(DEFAULT_HTTP_TIMEOUT_SECS);
+
+    reqwest::Client::builder()
+        .timeout(Duration::from_secs(http_timeout))
+        .build()
+        .context("Failed to create default HTTP client")
+}
+
+/// Create the default OCI client used when none is supplied.
+fn default_oci_client() -> Result<oci_client::Client> {
+    let oci_timeout = std::env::var("OCI_TIMEOUT_SECS")
+        .ok()
+        .and_then(|s| s.parse().ok())
+        .unwrap_or(DEFAULT_OCI_TIMEOUT_SECS);
+
+    Ok(oci_client::Client::new(oci_client::client::ClientConfig {
+        read_timeout: Some(Duration::from_secs(oci_timeout)),
+        ..Default::default()
+    }))
+}
+
+impl LifecycleConfig {
+    pub(crate) fn destructure(self) -> LifecycleConfigParts {
```

Comment:
I feel like this and the `LifecycleConfigParts` is redundant and weird to use. I would either make `into_parts` a `pub(crate)` or make each of the data members of `LifecycleConfig` `pub(crate)`